### ISSUE TITLE
Getting an error: NullInjectorError: No provider for OutletService!

### ISSUE THREAD

**Issue by LighthouseKrakenDSM** [MTobisch/ngx-dynamic-hooks#38]
When trying to include this package in an angular app, i get the following error:

**ERROR NullInjectorError: R3InjectorError(AppModule)[OutletService -> OutletService -> OutletService]: 
  NullInjectorError: No provider for OutletService!**

here is my app.module.ts

`import { BrowserModule } from '@angular/platform-browser';`
`import { NgModule} from '@angular/core';`
`import { AppRoutingModule } from './app-routing.module';`
`import { AppComponent } from './app.component';`
`import { HashLocationStrategy, LocationStrategy } from '@angular/common';`
`import { FormsModule, ReactiveFormsModule } from '@angular/forms';`
`import { DynamicHooksModule, HookParserEntry } from 'ngx-dynamic-hooks';`
`import { KrakenButtonModule, KrakenButtonComponent } from 'kraken-button';`

`const componentParsers: Array<HookParserEntry> = [`
 ` {component: KrakenButtonComponent}`
`];`

`@NgModule({`
`  declarations: [`
`    AppComponent,`
`],`
` imports: [`
`    BrowserModule,`
`    AppRoutingModule,`
`    DynamicHooksModule,`
`providers: [`
`    {provide: LocationStrategy, useClass: HashLocationStrategy}`
`  ],`
`  entryComponents: [],`
`  bootstrap: [AppComponent]`
`})`
`export class AppModule { `
`  constructor() {`

`  }`
`}`

And in my component html file i have: 

`<ngx-dynamic-hooks [content]="content" [context]="context" [options]="{sanitize: false}"></ngx-dynamic-hooks>`

My component ts file has:

`public content: any;`
`public context: any;`

What could be causing this error? I have previous added this package to a few component libraries with no problems, but this is the first time I'm trying to use it within an angular app. I'm running angular 15

**Comment by MTobisch** [comment]
Hello there,

is it possible that you just forgot to call `forRoot` on the module? Like so:

```
  imports: [
      BrowserModule,
      AppRoutingModule,
      DynamicHooksModule.forRoot({
        globalParsers: componentParsers
      }),
  ]
```
It would kind of explain the error, as without `forRoot`, the components would be imported, but not the services, resulting in the `ngx-dynamic-hooks`-component not finding `OutletService`.

**Comment by LighthouseKrakenDSM** [comment]
yeah, that was it, man, thanks so much! And also for the quick response. I was going to wrap up for the day, I figured I may get a response tomorrow lol. Now you kept me working!

**Comment by MTobisch** [comment]
Ha, no problems man. Always happy when its not a bug with the library 😄

**Comment by LighthouseKrakenDSM** [comment]
One more issue im running into. Im not able to use `[(ngModel)]`.  If I have `ngModel` on the a component that im passing into `ngx-dynamic-hooks [content]="content"`, the component will not render. Is that by design?

**Comment by MTobisch** [comment]
Are you using `[(ngModel)]` directly in the content string? Kind of like so:  `let content = "... <input [(ngModel)]="somevariable"> ..."`?

Because that unfortunately won't work. `ngModel` is ultimately a directive and the library does not parse directives, only selectors for loading components (hooks).

It should function however inside a component that was loaded by the library. If the content string is `let content = "<app-somecomponent></app-somecomponent>"` and in the template of SomeComponent you then use `[(ngModel)]`, it should work again.

**Comment by LighthouseKrakenDSM** [comment]
ok, that makes sense. So a component that has ngModel within it will work fine, it just wont work if its a part of the component wrapper like this `<my-component [(ngModel)]="myVar"></my-component>`

**Comment by MTobisch** [comment]
Yes, indeed.

Similarly, Angular template syntax like `*ngIf`, `*ngFor`, etc. also will not work **directly in the content string**. This is because the library simply takes that content string, looks for any component you (typically) registered via `forRoot()` and simply loads them in place of their hook/selector. It does nothing else. 

Loading directives dynamically is impossible last I checked and even if not, at that point, you'd be basically mimicing building a Angular template compiler. Not in the scope of this library.

However, in the components the library has loaded, you can then once again use any Angular magic that you want, just like with any other component in your app.
